docs/source/api/notebooks/Notebook_4_Customized_Models.ipynb

```diff
@@ -0,0 +1,456 @@
+    "\n",
+    "            # check if no edges exist for this edge type\n",
+    "            if sub_graph.num_edges() == 0:\n",
+    "                continue\n",
+    "\n",
+    "            # extract source and destination node features\n",
+    "            src_feat = inputs[src_type]\n",
+    "            dst_feat = inputs[dst_type][ :sub_graph.num_dst_nodes()]\n",
+    "\n",
+    "            # GAT in one relation type\n",
+    "            agg = self.rel_gats[str((src_type, e_type, dst_type))](sub_graph, (src_feat, dst_feat))\n",
+    "            agg = agg.view(agg.shape[0], -1)\n",
+    "\n",
+    "            # store aggregations in destination nodes\n",
+    "            sub_graph.dstdata['agg_' + str((src_type, e_type, dst_type))] = self.leaky_relu(agg)\n",
+    "\n",
+    "        h = {}\n",
+    "        for n_type in g.dsttypes:\n",
+    "            if g.num_dst_nodes(n_type) == 0:\n",
+    "                continue\n",
+    "\n",
+    "            # cross relation attention enhancement as outputs\n",
+    "            agg_list = []\n",
+    "            for k, v in g.dstnodes[n_type].data.items():\n",
+    "                if k.startswith('agg_'):\n",
+    "                    agg_list.append(v)\n",
+    "\n",
+    "            # cross-relation attention\n",
+    "            cr_agg = th.stack(agg_list, dim=1)\n",
+    "\n",
+    "            cr_att = th.matmul(cr_agg, self.cr_attn_weights)\n",
+    "            cr_sfm = th.softmax(cr_att, dim=1)\n",
+    "\n",
+    "            # cross-relation weighted aggregation\n",
+    "            cr_sum = (cr_agg * cr_sfm).sum(dim=1)\n",
+    "\n",
+    "            # process new features\n",
+    "            if self.self_loop:\n",
+    "                cr_sum = cr_sum + th.matmul(inputs[n_type][ :g.num_dst_nodes(n_type)], self.loop_weight)\n",
+    "            if self.bias:\n",
+    "                cr_sum = cr_sum + self.h_bias\n",
+    "            if self.activation:\n",
+    "                cr_sum = self.activation(cr_sum)\n",
+    "            if self.norm:\n",
+    "                cr_sum = self.norm[n_type](cr_sum)\n",
+    "            cr_sum = self.dropout(cr_sum)\n",
+    "\n",
+    "            h[n_type] = cr_sum\n",
+    "\n",
+    "        return h\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2.2 `Crae_GatEncoder` implementation\n",
+    "\n",
+    "Here we implement the `Crae_GatEncoder` by extending the base GraphStorm encoder, `graphstorm.model.gnn_encoder_base.GraphConvEncoder`, and implementing the `forward(self, blocks, h)` funciton to make this class compatible to GraphStorm model architecture. The `forward()` function takes a DGL block list and a dictionary of node representations as input argument, and return a dictionary that contains the new node representations. This `forward()` function will be called by `graphstorm.model.GSgnnNodeModel`on its `forward()` function.\n",
```

Comment:
```suggestion
    "Here we implement the `Crae_GatEncoder` by extending the base GraphStorm encoder, `graphstorm.model.gnn_encoder_base.GraphConvEncoder`, and implementing the `forward(self, blocks, h)` funciton as required by GraphConvEncoder. The `forward()` function takes a DGL block list and a dictionary of node representations as input arguments, and returns a dictionary that contains the new node representations. This `forward()` function will be called by `graphstorm.model.GSgnnNodeModel`on its `forward()` function.\n",
```

Reply:
Why GSgnnNodeModel specifically?

Reply:
1. `GraphConvEncoder` does not has a specific `forward()` API requirement. And all GNN encoders' `forward()` API requirements are **implicitly** defined in the `compute_embed_step()` method at the `graphstorm.model.GSgnnModel'.
2. `GSgnnNodeModel` extends the `GSgnnModel`. And its `forward()` function calls the `compute_embed_step()`.

So in order to explain this example code, here just gives a rough idea about who will call the `forward(self, blocks, h)` function.